### DISCUSSION TITLE
Webpack roadmap regarding module federation 2.0

### DISCUSSION THREAD

**Discussion by rodrigoBerlochi** [webpack/webpack#18462] [General]
Hello! Module Federation version 2.0 (https://module-federation.io/blog/announcement.html) has recently been released. Now that it is decoupled from Webpack, I wonder: Does the Webpack team plan to completely remove the internal plugin from Webpack's module federation, or will it still be available? And finally, if it is kept as a Webpack plugin, will it have the same updates as version 2.0 or will the implementation be the same as it is now?

Thanks!

**Comment by alexander-akait** [comment]
> Decoupling Webpack and Supporting Cross-Tool Sharing of Modules[#](https://module-federation.io/blog/announcement.html#decoupling-webpack-and-supporting-cross-tool-sharing-of-modules)
The new version of the Module Federation plugin adopts a completely new Runtime architecture. This change means that the build tool is no longer tightly coupled with the Runtime. Developers can now freely introduce artifacts produced by different build tools into their projects and flexibly assemble them for operation.

> Currently, multiple build tools are compatible with it, including Webpack and Rspack. Other build tools can also implement Module Federation based on this standard, which will maintain consistency and standardization in the runtime.

/cc @ScriptedAlchemy Do we need something special to improve/support v2?

**Reply by ScriptedAlchemy** [reply]
Alex this may require more discussions. We decoupled it from webpack to support rspack, esbuild, rollup. This also occurred during the time where webpack had a gap in maintainers who had merge powers were all unresponsive.

Depending on webpacks stance. We could do an implementation similar to how webpack/rspack work today with enhanced which is to initialize our runtime. 
A "major" issue for us is bugs in the semver parser used for shared. Versions with ~ are not respected and we could not figure out the regexs to fix it in semver.js 

This also contributed about 20kb to the bundle size since we have to use a npm package to support ~ which accounts for the majority of the size increase in v2 vs v1 

V2 also had to patch multiple issues that federation has had with webpack some not as elegant as id like but best we could do. (Single runtime chunk, split chunks for example)

I would love to minimize the confusion / fragmentation between them.

How would you see this being implemented in webpack? Or do you want to discuss this in more depth via slack? 

V2 isn't perfect but without webpacks help we could not change its core functionality. Rspack inherited some of this as well but we are able to change and release it on demand since we own the whole supply chain. 

One thing I'd prefer to preserve is the iteration speeds we can achieve with federation as a standalone library. 
Webpack has a high risk exposure for obvious reasons so I do understand the rationale behind it. 

Outside the core however it only took us a month or two to rewrite it and we drastically expand its capabilities each month since. 

I think we would also need to consider the scope. 

Enhanced has typescript remote type hinting, chrome tools, hmr support, json version of the remotes etc - from my years with webpack - I believe this is counterintuitive to the teams preference to keep the repo core small.

**Reply by ScriptedAlchemy** [reply]
Update: i went to find the semver issue and looks like it was fixed 2 days ago at long last. https://github.com/webpack/webpack/issues/17756

This will greatly reduce the overhead on our end, i will look to remove the semver package again and port webpacks back into my package.

**Reply by alexander-akait** [reply]
@ScriptedAlchemy Maybe we need to merge something else to help? Feel free to point me

**Reply by rodrigoBerlochi** [reply]
Thanks to both of you! I have now a better context. 🙇
